### PR TITLE
Add support for Raspberry Pi boards

### PR DESCRIPTION
**Pull request by mirzak**
This pull request adds support to build mender for Raspberry Pi boards. I have only tested this on a Raspberry Pi 2.

This adds a dependency on (when building for a Raspberry Pi board):
```
URI: git://git.yoctoproject.org/meta-raspberrypi
branch: master
revision: HEAD
```
Following needs to be added to local.conf when building for Raspberry Pi`s:
```
MACHINE ?= "raspberrypi2"

KERNEL_IMAGETYPE = "uImage"

INHERIT += "mender-full"

DISTRO_FEATURES_append = " systemd"
VIRTUAL-RUNTIME_init_manager = "systemd"
DISTRO_FEATURES_BACKFILL_CONSIDERED = "sysvinit"
VIRTUAL-RUNTIME_initscripts = ""
IMAGE_FSTYPES = "ext4"

MENDER_ARTIFACT_NAME = "my-mender-image-1.0"
MENDER_PARTITION_ALIGNMENT_MB = "4"
MENDER_BOOT_PART_SIZE_MB = "40"

MACHINE_ESSENTIAL_EXTRA_RDEPENDS_append = " bcm2835-bootfiles"

# raspberrypi files aligned with mender layout requirements
IMAGE_BOOT_FILES_append = " boot.scr u-boot.bin;${SDIMG_KERNELIMAGE}"
IMAGE_INSTALL_append = " kernel-image kernel-devicetree"
```

### CODE REVIEW

**Comment by kacf** [comment]
Awesome! I'll take a look shortly!

File: meta-mender-core/classes/mender-sdimg.bbclass [hunk 1]
```diff
@@ -97,7 +101,7 @@ python() {
             d.setVar('IMAGE_TYPEDEP_sdimg_append', " " + fs)
 }
 
-IMAGE_DEPENDS_sdimg = "${IMAGE_DEPENDS_wic} dosfstools-native mtools-native"
+IMAGE_DEPENDS_sdimg = "${MENDER_IMAGE_EXTRA_DEPENDS} ${IMAGE_DEPENDS_wic} dosfstools-native mtools-native"
```
Comment (kacf) [inline meta-mender-core/classes/mender-sdimg.bbclass]: I believe you can get the same effect without adding this, by adding this to `local.conf`:

```
MACHINE_ESSENTIAL_EXTRA_RDEPENDS_append = " bcm2835-bootfiles"
```

Comment (mirzak) [inline meta-mender-core/classes/mender-sdimg.bbclass]: Makes sense. Will drop that patch.

Thank you for you review.

**Comment by mirzak** [comment]
Was a bit fast accepting that change :).

Build fails when I use

```
MACHINE_ESSENTIAL_EXTRA_RDEPENDS_append = " bcm2835-bootfiles"
```

Will need to look in to this a bit

**Comment by bboozzoo** [comment]
I would start by adding this to your `local.conf`:
```
EXTRA_IMAGEDEPENDS += " bcm2835-bootfiles"
```

I suppose this should actually be set by machine conf. From what I can see, `meta-raspberrypi` does not set it, so may want to submit a patch upstream at some point.

**Comment by kacf** [comment]
> `EXTRA_IMAGEDEPENDS += "bcm2835-bootfiles"`

At least for me this one didn't work, because it ended up building it too late, after the sdimg stage.

**Comment by mirzak** [comment]
Yes, got the same problem with EXTRA_IMAGEDEPENDS.

Following works and is the equivalent of my initial patch

```
IMAGE_DEPENDS_sdimg += " bcm2835-bootfiles"
```

Acceptable?

**Comment by kacf** [comment]
I'm good with that! Maybe this, and the other `local.conf` modifications you have, are good candidates for a `README.md` file in the `meta-mender-raspberry` folder, what do you think?

**Comment by mirzak** [comment]
> I'm good with that! Maybe this, and the other local.conf modifications you have, are good candidates for a README.md file in the meta-mender-raspberry folder, what do you think?

+1 on that. I will create one and push to this pull request.

**Comment by mirzak** [comment]
> Yes, got the same problem with EXTRA_IMAGEDEPENDS.

> Following works and is the equivalent of my initial patch

> IMAGE_DEPENDS_sdimg += " bcm2835-bootfiles"
> Acceptable?

Did a full re-build. Above does not work, I was fooled by "old" build state.

To start from the beginning. Ultimately I would like to add a dependency to "sdimg". Which my initial patch did and I think that this is the only "correct" way (not maybe how I did it, but I need a way to add this dependency). At the moment I can not see how I could do that.

```
 IMAGE_DEPENDS_sdimg += " bcm2835-bootfiles"
```
is overwritten by
```
IMAGE_DEPENDS_sdimg = "${IMAGE_DEPENDS_wic} dosfstools-native mtools-native"
```
in the mender-sdimg.bbclass.

What bcm2835 provides are some binary files which are in IMAGE_BOOT_FILES, and hence the only dependency is "sdimg/wic". To add this dependency anywhere else seems wrong and produces unwanted behavior.

Something that would fix my problem here is changing:

```
IMAGE_DEPENDS_sdimg = "${IMAGE_DEPENDS_wic} dosfstools-native mtools-native"
```
to

```
IMAGE_DEPENDS_sdimg += "${IMAGE_DEPENDS_wic} dosfstools-native mtools-native"
```

Then it wont overwrite my change in local.conf.

Thoughts? Tips?

**Comment by kacf** [comment]
I think that's alright, changing `=` to `+=`.

**Comment by mirzak** [comment]
Any comments on the updates?

**Comment by kacf** [comment]
And in it goes! Thanks a lot!

**Pull request merged**